php/php-rr-sdk.md

```diff
@@ -0,0 +1,692 @@
+
+#### Timers
+Workflow timers can operate as promise or blocking operation:
+
+```php
+$timer = yield $this->timer(1 * Time::DAY);
+
+yield $timer->wait();
+```
+
+Blocking:
+
+```php
+yield $this->timer(1 * Time::DAY)->wait();
+
+// or
+yield $this->sleep(1 * Time::DAY);
+```
+
+Can be comfined with Promises:
+
+```php
+$a = yield $this->activities->doSomething(...);
+$t = yield $this->timer(1 * Time::MINUTE);
+
+yield $this->waitAny($a, $t);
+
+if ($t->isFired()) {
+    // ...
+}
+```
+
+### Queries
+Workflow queries can be easily implemented using dedicated query method of workflow. The state of the workflow can be 
+extracted from the properties of the object.
+
+```php
+use Temporal\Workflow;
+
+class DemoWorkflow extends Workflow\Workflow
+{
+    private $step = 0;
+
+    public function queryStep(): int 
+    {
+        return $this->step;
+    }
+
+    public function run(string $input): string
+    {
+        yield $this->activities->doSomething($input);
+        $this->step++;
+
+        // ...
+    }
+}
+```
+
+> Query methods must be registered explicitly.
+
+An alternative approach can be based on registering the callback function during the workflow execution:
+
+```php
+yield new QueryHandler('something', function() use (&$step) {
+    return $step;
+});
+```
+
+> Such an approach is more suitable for dynamic workflows. The principle of work is identical to pre-registered queries.
+
+#### Using Annotations
+For more advanced frameworks it should be possible to register query method using annotation or attribute (PHP8) as 
+in Java SDK.
+
+```php
+/**
+ * @Workflow\QueryMethod(name="something")
+ */
+public function querySomething(string $world): string 
+{
+    return 'hello';
+}
+```
+
+### Signals
+Unlike queries the signals are fist-class citizen of actual workflow code. Unlike activities the signal can be triggered
+at any moment, without prior expectation by workflow. 
+
+Signal method can be defined explicitly:
+
+```php
+class UploadWorkflow extends Workflow\Workflow
+{
+    /** @Temporal\SignalMethod(name="signal") */
+    public function handleSignal($input):void
+    {
+    
+    }
+
+    public function run(string $input): string
+    {
+        
+    }
+}
+```
+
+Or at runtime:
+
+```php
+public function run(string $input): string
+{
+    yield new SignalHandler('name', function($input){
+    
+    });       
+}
+```
+
+> Need Temporal authors feedback.
+
+#### Handle Signals
+Incoming signals can be handled during any workflow `yield` call. Workflow should allow to explicitly wait for signal:
+
+```php
+public function run(string $input): string
+{
+    yield new SignalHandler('name', function($input){
+        // do something
+    });       
+
+    yield new WaitSignal('name');
+}
+```
+
+### Sessions
+Sessions can be created as sub-activity group, using similar approach as `withTimeout` and other methods. However,
+the explicit yield required. 
```

Comment:
An alternative approach is used by Go client is to associate a session with ctx object. Then all other operations that reference the session context are considered part of a session.

Reply:
See above. All `yield` commands associated with $ctx by default. So in this case we are altering the default values of generated (`yield` ed) commands.  The problem with sessions is that we have to explicitly open it via command, so we can't use `with` but rather `yield`.